// @expo/dev-server/build/MetroDevServer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDevServerMiddleware = exports.LogReporter = exports.attachInspectorProxy = exports.bundleAsync = exports.runMetroDevServerAsync = void 0;
const config_1 = require("@expo/config");
const ExpoMetroConfig = __importStar(require("@expo/metro-config"));
const chalk_1 = __importDefault(require("chalk"));
const HermesBundler_1 = require("./HermesBundler");
const LogReporter_1 = __importDefault(require("./LogReporter"));
exports.LogReporter = LogReporter_1.default;
const createDevServerAsync_1 = require("./metro/createDevServerAsync");
const importMetroFromProject_1 = require("./metro/importMetroFromProject");
const devServerMiddleware_1 = require("./middleware/devServerMiddleware");
Object.defineProperty(exports, "createDevServerMiddleware", { enumerable: true, get: function () { return devServerMiddleware_1.createDevServerMiddleware; } });
async function runMetroDevServerAsync(projectRoot, options) {
    const reporter = new LogReporter_1.default(options.logger);
    const metroConfig = await ExpoMetroConfig.loadAsync(projectRoot, { reporter, ...options });
    const { middleware, attachToServer } = devServerMiddleware_1.createDevServerMiddleware({
        port: metroConfig.server.port,
        watchFolders: metroConfig.watchFolders,
        logger: options.logger,
    });
    const customEnhanceMiddleware = metroConfig.server.enhanceMiddleware;
    // @ts-ignore can't mutate readonly config
    metroConfig.server.enhanceMiddleware = (metroMiddleware, server) => {
        if (customEnhanceMiddleware) {
            metroMiddleware = customEnhanceMiddleware(metroMiddleware, server);
        }
        return middleware.use(metroMiddleware);
    };
    const { server } = await createDevServerAsync_1.createDevServerAsync(projectRoot, {
        config: metroConfig,
        logger: options.logger,
    });
    const { messageSocket, eventsSocket } = attachToServer(server);
    reporter.reportEvent = eventsSocket.reportEvent;
    return {
        server,
        middleware,
        messageSocket,
    };
}
exports.runMetroDevServerAsync = runMetroDevServerAsync;
let nextBuildID = 0;
// TODO: deprecate options.target
async function bundleAsync(projectRoot, expoConfig, options, bundles) {
    const metro = importMetroFromProject_1.importMetroFromProject(projectRoot);
    const Server = importMetroFromProject_1.importMetroServerFromProject(projectRoot);
    const reporter = new LogReporter_1.default(options.logger);
    const config = await ExpoMetroConfig.loadAsync(projectRoot, { reporter, ...options });
    const buildID = `bundle_${nextBuildID++}`;
    const metroServer = await metro.runMetro(config, {
        watch: false,
    });
    const buildAsync = async (bundle) => {
        var _a, _b, _c, _d;
        const bundleOptions = {
            ...Server.DEFAULT_BUNDLE_OPTIONS,
            bundleType: 'bundle',
            platform: bundle.platform,
            entryFile: bundle.entryPoint,
            dev: (_a = bundle.dev) !== null && _a !== void 0 ? _a : false,
            minify: (_b = bundle.minify) !== null && _b !== void 0 ? _b : !bundle.dev,
            inlineSourceMap: false,
            sourceMapUrl: bundle.sourceMapUrl,
            createModuleIdFactory: config.serializer.createModuleIdFactory,
            onProgress: (transformedFileCount, totalFileCount) => {
                if (!options.quiet) {
                    reporter.update({
                        buildID,
                        type: 'bundle_transform_progressed',
                        transformedFileCount,
                        totalFileCount,
                    });
                }
            },
        };
        reporter.update({
            buildID,
            type: 'bundle_build_started',
            bundleDetails: {
                bundleType: bundleOptions.bundleType,
                platform: bundle.platform,
                entryFile: bundle.entryPoint,
                dev: (_c = bundle.dev) !== null && _c !== void 0 ? _c : false,
                minify: (_d = bundle.minify) !== null && _d !== void 0 ? _d : false,
            },
        });
        const { code, map } = await metroServer.build(bundleOptions);
        const assets = (await metroServer.getAssets(bundleOptions));
        reporter.update({
            buildID,
            type: 'bundle_build_done',
        });
        return { code, map, assets };
    };
    const maybeAddHermesBundleAsync = async (bundle, bundleOutput) => {
        var _a, _b;
        const { platform } = bundle;
        const isHermesManaged = HermesBundler_1.isEnableHermesManaged(expoConfig, platform);
        const paths = config_1.getConfigFilePaths(projectRoot);
        const configFilePath = (_b = (_a = paths.dynamicConfigPath) !== null && _a !== void 0 ? _a : paths.staticConfigPath) !== null && _b !== void 0 ? _b : 'app.json';
        await HermesBundler_1.maybeThrowFromInconsistentEngineAsync(projectRoot, configFilePath, platform, isHermesManaged);
        if (isHermesManaged) {
            const platformTag = chalk_1.default.bold({ ios: 'iOS', android: 'Android', web: 'Web' }[platform] || platform);
            options.logger.info({ tag: 'expo' }, `💿 ${platformTag} Building Hermes bytecode for the bundle`);
            const hermesBundleOutput = await HermesBundler_1.buildHermesBundleAsync(projectRoot, bundleOutput.code, bundleOutput.map, bundle.minify);
            bundleOutput.hermesBytecodeBundle = hermesBundleOutput.hbc;
            bundleOutput.hermesSourcemap = hermesBundleOutput.sourcemap;
        }
        return bundleOutput;
    };
    try {
        const intermediateOutputs = await Promise.all(bundles.map(bundle => buildAsync(bundle)));
        const bundleOutputs = [];
        for (let i = 0; i < bundles.length; ++i) {
            // hermesc does not support parallel building even we spawn processes.
            // we should build them sequentially.
            bundleOutputs.push(await maybeAddHermesBundleAsync(bundles[i], intermediateOutputs[i]));
        }
        return bundleOutputs;
    }
    finally {
        metroServer.end();
    }
}
exports.bundleAsync = bundleAsync;
/**
 * Attach the inspector proxy to a development server.
 * Inspector proxy is used for viewing the JS context in a browser.
 * This must be attached after the server is listening.
 * Attaching consists of pushing custom middleware and appending WebSockets to the server.
 *
 *
 * @param projectRoot
 * @param props.server dev server to add WebSockets to
 * @param props.middleware dev server middleware to add extra middleware to
 */
function attachInspectorProxy(projectRoot, { server, middleware }) {
    const { InspectorProxy } = importMetroFromProject_1.importInspectorProxyServerFromProject(projectRoot);
    const inspectorProxy = new InspectorProxy(projectRoot);
    if ('addWebSocketListener' in inspectorProxy) {
        // metro@0.59.0
        inspectorProxy.addWebSocketListener(server);
    }
    else if ('createWebSocketListeners' in inspectorProxy) {
        // metro@0.66.0
        // TODO: This isn't properly support without a ws router.
        inspectorProxy.createWebSocketListeners(server);
    }
    // TODO(hypuk): Refactor inspectorProxy.processRequest into separate request handlers
    // so that we could provide routes (/json/list and /json/version) here.
    // Currently this causes Metro to give warning about T31407894.
    // $FlowFixMe[method-unbinding] added when improving typing for this parameters
    middleware.use(inspectorProxy.processRequest.bind(inspectorProxy));
    return { inspectorProxy };
}
exports.attachInspectorProxy = attachInspectorProxy;
